state/self_check.go

```diff
@@ -0,0 +1,262 @@
+package state
+
+import (
+	"fmt"
+	"sync"
+)
+
+// The self-check mechanism is intended to catch errors and potenttial problems
+// encountered during a test surface them to users in a test summary, and
+// communicate how these errors will impact the various features of pganalyze.
+
+type CollectionStateCode int
+
+const (
+	CollectionStateUnchecked CollectionStateCode = iota
+	CollectionStateNotAvailable
+	CollectionStateWarning
+	CollectionStateError
+	CollectionStateOkay
+)
+
+type CollectionAspectStatus struct {
+	State CollectionStateCode
+	Msg   string
+	Hint  string
+}
+
+type CollectionAspect int
+
+const (
+	CollectionAspectTelemetry CollectionAspect = iota
+	CollectionAspectSystemStats
+	CollectionAspectMonitoringDbConnection
+	CollectionAspectPgVersion
+	CollectionAspectApiConnection
+	CollectionAspectPgStatStatements
+	CollectionAspectLogs
+	CollectionAspectExplain
+)
+
+type DbCollectionAspect int
+
+const (
+	CollectionAspectSchemaInformation DbCollectionAspect = iota
+	CollectionAspectColumnStats
+	CollectionAspectExtendedStats
+)
+
+type SelfCheckStatus struct {
+	mutex               *sync.Mutex
+	CollectionSuspended struct {
+		Value bool
+		Msg   string
+	}
+	MonitoredDbs     []string
+	AspectStatuses   map[CollectionAspect]*CollectionAspectStatus
+	AspectDbStatuses map[DbCollectionAspect](map[string]*CollectionAspectStatus)
+}
+
+func MakeSelfCheck() (s *SelfCheckStatus) {
+	return &SelfCheckStatus{
+		mutex:            &sync.Mutex{},
+		AspectStatuses:   make(map[CollectionAspect]*CollectionAspectStatus),
+		AspectDbStatuses: make(map[DbCollectionAspect](map[string]*CollectionAspectStatus)),
+	}
+}
+
+// collection suspended (e.g., if replica)
+func (s *SelfCheckStatus) MarkCollectionSuspended(format string, args ...any) {
+	if s == nil {
+		return
+	}
```

Comment:
Checking for nil like this means the method can be invoked on a nil value of the right type, so we can easily disable the self-check without having to guard on `options.TestRun` everywhere.